### PR TITLE
Adds terraform 0.12 support

### PR DESCRIPTION
**Pull request by userhas404d**
Fixes https://github.com/dependabot/dependabot-core/issues/1176

I opted for both [hcl2json](https://github.com/tmccombs/hcl2json) and [terraform-config-inspect](https://github.com/hashicorp/terraform-config-inspect). 

`hcl2json` for terragrunt and 
`terraform-config-inspect` for tf 0.12

I wanted to go with `terraform-config-inspect` for both, but it didn't play nice with terragrunt :disappointed:

Notes: 
- I created my own fork of [hcl2json](https://github.com/userhas404d/hcl2json/releases) in order to publicly host the binaries for it
- opting for `terraform-config-inspect` adds go as a dependency to being able to support this language. Not sure if that's kosher or not..

### CODE REVIEW

**Comment by edahlseng** [comment]
Is there anything we can do to help push this through?

**Comment by hmarr** [comment]
Thanks for submitting this, and sorry we've been so slow to get to it. Honestly, the thing that would help us most at the minute would be filling [this role](https://boards.greenhouse.io/github/jobs/1862178), so if you know someone suitable, please  encourage them to apply!

Until we have that role filled, we're unable to dedicate much time to languages not [supported by the dependency graph](https://help.github.com/en/articles/listing-the-packages-that-a-repository-depends-on#supported-languages). Longer term, we do have a good plan for supporting a wide range of languages, and making it much easier for the community to help with languages that aren't officially supported, so please bear with us during this slightly awkward transition period. In the meantime, I'll see if I can get this reviewed over the weekend or one evening in the next few days.

At a first glance, the main thing that stands out to me is downloading the binary dependency from the fork of hcl2json. In particular, there's currently not much visibility over the contents of that binary. Two options I see are:

1. Vendor the dependency, or clone it and pin to a specific commit in the build script. Then we can build it in the build script, avoiding the opaque binary dependency.
2. We could fork it into the dependabot org and run the build in a GitHub Action, where the build and release logs would be visible, and we'd have control over the release process.

Let me know what you think. And once again, sorry for taking so long to get to this!

**Comment by userhas404d** [comment]
@hmarr completely understand, and glad to hear that there's a plan forward with enhanced language support as well. 

In regards to your concerns regarding hcl2json:  

I've opened a [PR](https://github.com/tmccombs/hcl2json/pull/4) upstream but it looks like @tmccombs is busy as well. My initial recommendation would be for option 2. Although, based on the dependabot team's time/resource constraints that sounds like going that route will take longer? Additionally, if the plan is to go that route with hcl2json does the same need to be done with json2hcl?

**Comment by rebelagentm** [comment]
It was mentioned [here](https://github.com/dependabot/feedback/issues/671#issuecomment-538021023) that HCL 2 is no longer experimental.  I'm not sure if that affects this PR, but FYI just in case.

**Comment by hairyhenderson** [comment]
> Honestly, the thing that would help us most at the minute would be filling [this role](https://boards.greenhouse.io/github/jobs/1862178)

@hmarr Looks like that role is filled now (or at least, is showing as "no longer open") - is there any possibility on 0.12 support moving forward?

**Comment by lorengordon** [comment]
@dependabot @hmarr @feelepxyz anything we can do here to move this forward?

**Comment by feelepxyz** [comment]
@lorengordon we've paused dependabot-core improvements while we focus on integrating Dependabot natively in GitHub, this is still a few months out, unfortunately.

File: terraform/lib/dependabot/terraform/file_parser.rb [hunk 1]
```diff
@@ -218,13 +241,43 @@ def source_type(source_string)
...
+          )
+        rescue SharedHelpers::HelperSubprocessFailed => e
+          @msg = "hcl2json: " + e.message.strip + "\n"
+          begin
+            parsed_file_helper(
+              file,
+              # terraform-config-inspect only works when provided a directory
+              "terraform-config-inspect --json .",
+              # specifying $GOPATH/bin:$PATH here doesn't seem to work..
+              "PATH" => "/opt/go/gopath/bin:$PATH"
```
Comment (patrickjahns) [inline terraform/lib/dependabot/terraform/file_parser.rb:265]: I've experienced issues with this when running on osx, as I did not specify GOPATH and thus the modules where at `/Users/username/go/bin` 

I Would suggest to use `terraform_parser_path` here as well and rather modify the helper script to copy the binary to the correct location

**Comment by patrickjahns** [comment]
I've given this a local test and it works - some minor remarks:

- hardcoding the gopath for `terraform-config-inspect` has caused me a bit of pain 
- fetching `terraform-config-inspect` without specifying `GO111MODULE=on` was not working (see upstream issue https://github.com/hashicorp/terraform-config-inspect/issues/36 )

About `terraform-config-inspect` - I believe it would make sense to also pre-built the biaries here and host them, so we need to only fetch them as with other dependencies. Solves quite a bit of headache for setting this up - what do you think @userhas404d 

Other than that - thanks for doing this - thinking now that I will create a temporary github action with this functionality

**Comment by patrickjahns** [comment]
I've created a terraform action to provide the features from this pull request for the time being.
https://github.com/marketplace/actions/terraform-dependabot

Thanks to @userhas404d for the great initial work

**Comment by userhas404d** [comment]
thanks for the feedback @patrickjahns! I've since implemented some of the updates you suggested and a few others that I found after running this implementation against my org's terraform/terragrunt private projects. I'll open a PR against your github action here shortly. 

re GOPATH/`terraform_config_inspect`: I'll have to defer how to handle that to the dependabot-core maintainers. Earlier on @hmarr had some concerns over how to manage the other pre-compile bins being used in this solution, and I'm not sure where that stands.  There are a few ([1](https://github.com/hashicorp/terraform-config-inspect/issues/30), [2](https://github.com/hashicorp/terraform-config-inspect/issues/41)) long-ish running issues open against the `terraform_config_inspect`  project requesting that automated releases be provided, but until those are addressed it looks like compiling it in the terraform helper is probably the best way forward.

**Comment by userhas404d** [comment]
I'm adding this comment here as a reminder to myself (or anyone else who want's to take a stab at fixing it): Hit a bug with this implementation supporting both legacy and current versions of terragrunt config where dependabot will bomb when trying to parse a tf012 `.tfvars` as a legacy terragrunt file. Thinking that the best way forward (while maintaining backwards compatibility) will be a try/catch block around legacy terragrunt config processing.

**Comment by lorengordon** [comment]
> we've paused dependabot-core improvements while we focus on integrating Dependabot natively in GitHub, this is still a few months out, unfortunately.

Hi @feelepxyz just thought I'd follow up now that it's been a few months since the last update. Any word on resuming improvements like this? We've been using the github-action based on this branch for quite a while now and it's working well.

**Comment by chenrui333** [comment]
Considering now we have terraform 0.13 out, where are we with this support? :)

**Comment by brondum** [comment]
Any updates on this one ?

**Comment by jurre** [comment]
Terraform 12 (HCL 2) support is now available, thank you so much for this PR, it was instrumental in landing this.

**Pull request closed without merging**